src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,263 @@
+      inputActive.current?.blur();
+
+      return;
+    }
+
+    if (!trimmedTitle) {
+      deleteTodos(todo.id)
+        .then(() => {
+          handleTodoDestroy(todo);
+        })
+        .catch(() => {
+          handleError(Errors.deletingError);
+        })
+        .finally(() => {
+          dispatch({ type: 'setProcessingList', payload: [] });
+        });
+
+      return;
+    }
+
+    updateTodo(id, { ...todo, title: editedTitle.trim() })
+      .then(() => {
+        const newTodos = todos.map(currentTodo =>
+          currentTodo.id === id
+            ? { ...currentTodo, title: editedTitle.trim() }
+            : currentTodo,
+        );
+
+        dispatch({ type: 'setTodos', payload: newTodos });
+      })
+      .catch(() => {
+        handleError(Errors.updateError);
+        isError = true;
+      })
+      .finally(() => {
+        inputActive.current?.blur();
+
+        if (isError) {
+          inputActive.current?.focus();
+          setEditedTitle(editedTitle.trim());
+        }
+
+        dispatch({ type: 'setProcessingList', payload: [] });
+      });
+  };
+
+  const handleBlur = (
+    e: React.ChangeEvent<HTMLInputElement>,
+    handleTodo: Todo,
+  ) => {
+    const changedTitleOnBlur = todos.map(elem => {
+      if (elem.id === handleTodo.id && e.target.value.length > 0) {
+        return {
+          ...elem,
+          title: e.target.value.trim(),
+        };
+      }
+
+      return elem;
+    });
+
+    if (processingList.length === 0) {
+      updateTitle();
+
+      dispatch({
+        type: 'setIsDoubleClicked',
+        payload: { state: false, id: null },
+      });
+      dispatch({ type: 'setTodos', payload: changedTitleOnBlur });
+    }
+  };
+
+  const submit = (e: React.FormEvent) => {
+    e.preventDefault();
+
+    updateTitle();
+  };
+
+  useEffect(() => {
+    if (inputActive.current && todo) {
+      inputActive.current.focus();
+    }
+  }, [isDoubleClicked.id, todos, todo]);
+
+  return (
+    <div
+      data-cy="Todo"
+      className={classNames('todo', {
+        completed,
+      })}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={handleChangeStatus}
+        />
+      </label>
+
+      {isDoubleClicked.id === todo.id ? (
+        <form onSubmit={submit}>
+          <input
+            ref={inputActive}
+            data-cy="TodoTitleField"
+            defaultValue={title}
+            type="text"
+            onBlur={e => handleBlur(e, todo)}
+            className="todo__title-field"
+            placeholder="Empty todo will be deleted"
+            onKeyUp={event => {
+              if (event.key === 'Escape') {
+                dispatch({
+                  type: 'setIsDoubleClicked',
+                  payload: { state: false, id: null },
+                });
+              }
+            }}
```

Comment:
Move into separate handler instead of anonymous function in case logic takes more than 1 line